### ISSUE TITLE
Promise/Callback Result Typings in VisualRecognitionV3

### ISSUE THREAD

**Issue by evanshortiss** [watson-developer-cloud/node-sdk#896]
## Issue

*TLDR; Inconsistent typings in Promises vs. Callbacks due to the `return_response` parameter. It would be nice if both return `Response<T>` interface to remove the `return_response` parameter. The only affect on developers is they would need to call `res.result` to get the JSON response from the API call instead of `res` itself being the JSON*

I've been working with the SDK in TypeScript (great choice on using TS for this SDK btw 😎) and was thrown off a bit by the typings around Promises.

When calling `getClassifier` on a `VisualRecognitionV3` instance I did the following and saw the return type was `any` instead of the correct return type as seen with callbacks.

```ts
// vr3.getClassifier returns a Promise<any> instead of the correct
// VisualRecognitionV3.Classifier type
const myClassifier = await vr3.getClassifier({
  classifier_id: myClassifierId
})
```

A workaround is to set the type manually like so:

```ts
const myClassifier: VisualRecognitionV3.Classifier = await vr3.getClassifier({
  classifier_id: myClassifierId
})
```

That didn't take long to figure out, but feels unintuitive. After digging into the code I think it's a result of blocks of code based on the `return_response` parameter that can be passed into calls. These look like so:

```ts
    if (!_callback) {
      return new Promise((resolve, reject) => {
        this.getClassifier(params, (err, bod, res) => {
          err ? reject(err) : _params.return_response ? resolve(res) : resolve(bod);
        });
      });
    }
```

From what I can see `return_response` is only supported if using Promises, but I haven't tested that so don't quote me 😃

Ultimately for Promise users the return type of each method on `VisualRecognitionV3` instances right now is something along the lines of `Promise<T | VisualRecognitionV3.Response<T>>` depending on the value they passed for `return_response`.

## Propsoal

To create better typings remove support for `return_response` and make all responses return a consistent format of `VisualRecognitionV3.Response<T>`, e.g:

Instead of:

```ts
// Definition
public getClassifier(params: VisualRecognitionV3.GetClassifierParams, callback?: VisualRecognitionV3.Callback<VisualRecognitionV3.Classifier>): Promise<any> | void

// Usage without return_response
const mc1: VisualRecognitionV3.Classifier = await vr3.getClassifier({
  classifier_id: myClassifierId,
})

mc1.classifier_id // JSON returned directly

// Usage (without return_response)
const mc2: VisualRecognitionV3.Response<VisualRecognitionV3.Classifier> = await vr3.getClassifier({
  classifier_id: IBM_CLASSIFIER_ID,
  return_response: true
})

mc2.result.classifier_id // access the JSON via result/data
```

We'd have just a single return type and it would be inferred correctly:

```ts
// Definition
public getClassifier(params: VisualRecognitionV3.GetClassifierParams, callback?: VisualRecognitionV3.Callback<VisualRecognitionV3.Response<VisualRecognitionV3.Classifier>>): Promise<VisualRecognitionV3.Response<VisualRecognitionV3.Classifier>> | void

// Usage
const myClassifier = await vr3.getClassifier({
  classifier_id: IBM_CLASSIFIER_ID,
})

mc2.result.classifier_id // access the JSON via result/data
```

**Comment by dpopp07** [comment]
@evanshortiss Thanks for the issue / proposal!

I very much agree with everything you have stated. The `return_response` parameter is something we haven't been happy with and we are planning on removing it and returning a single Response object as you suggest.

The problem is, this would constitute a breaking change and therefore we have to wait until our next major release which is scheduled for September (we coordinate major releases of all the Watson SDKs every 6 months).

I've added [an issue for tracking v5 goals](https://github.com/watson-developer-cloud/node-sdk/issues/897) and linked this issue to it.

Outside of just this issue, we are planning on making TypeScript-specific improvements to the SDK in that release. As a TypeScript user, I would be interested to hear any feedback you have on the TypeScript experience of this SDK so that we can improve it. Please leave any feedback in [the v5 issue](https://github.com/watson-developer-cloud/node-sdk/issues/897) or in a new issue. Thanks again!

**Comment by evanshortiss** [comment]
Thanks @dpopp07, will do! Understood on the major release, and good to know this is on the roadmap 😄

**Comment by dpopp07** [comment]
@evanshortiss I am working on this now. I noticed that the union type of `Promise` and `void` doesn't allow a user to access the `.then()` property of the Promise in TypeScript code. This applies to the current state of the code, so there is currently a bug. I'm guessing you haven't seen the error because you are using `async/await`. Currently, the only solution that I can think of that works for all scenarios is adding a third union of `any` to allow for both `Promise` and `void` behaviors. But this kind of defeats the purpose of the typing in the first place. Do you have any thoughts or comments about this issue? I'd love to improve this part of the TypeScript experience but the duality of returning Promises or using callbacks may be too big a limitation from a TS perspective

**Comment by evanshortiss** [comment]
@dpopp07, I see what you mean. I was using `async/await` so you're right in saying I never encountered the issue. 

Would simply returning a Promise everytime be a workable solution? It's a little wasteful since it would be creating a promise that _might_ not be used, but you can support callbacks, promises, and type both of them correctly AFAICT. Using method overloads should do the trick.

Here's a diff that shows how it works - [diff link](https://github.com/evanshortiss/node-sdk/compare/master..evanshortiss:promise-cb-overload-fix)

Screenshots with intellisense showing correctly typed `res` to the Promise below:

<img width="1440" alt="Screenshot 2019-09-20 at 5 43 29 PM" src="https://user-images.githubusercontent.com/1303687/65343963-43474180-dbce-11e9-9dd8-45a726967eb7.png">
<img width="1440" alt="Screenshot 2019-09-20 at 5 43 38 PM" src="https://user-images.githubusercontent.com/1303687/65343965-43474180-dbce-11e9-8d9f-66b6b7e4e248.png">

**Comment by evanshortiss** [comment]
Oh, and I didn't test this so the method body itself might need some more tweaks but it's a starting point 😄

**Comment by dpopp07** [comment]
@evanshortiss Thanks for the reply! I was just chatting about this with @MasterOdin and he had a similar solution. After going through it in detail, I think that's what I'll be implementing and you should get an inferred type 👍

**Comment by dpopp07** [comment]
@evanshortiss @MasterOdin 

Well, that didn't really work. The issue is, executing a callback function and later returning a Promise exposes the method to any problems that may exist in the function passed in as the callback. If the callback function throws an error and we end up returning a Promise, that Promise will reject. If the user thinks they are using callbacks, returning a silent Promise that resolves causes no harm. If the Promise rejects, however, they are left with an unhandled rejection error and wondering where in the world the Promise came from. This is certainly harmful.

After quite a bit of exploration for a solution, we essentially determined that callbacks and Promises just probably shouldn't mix - and that callbacks, Promises, and TypeScript certainly shouldn't mix.

The solution we decided on is to generate the correct type for the returned Promise union that with `any` to allow for the callback situation. It won't provide an inferred type and it is certainly not ideal. The plan moving forward is to try and remove support for callbacks completely in the future, possibly in v6, and solely use Promises. Unfortunately, we can't do this yet.

Let me know if you have any questions

**Comment by MasterOdin** [comment]
From my understanding, the worry here is that if `this.createRequest(parameters, _callback)` throws an exception while using a callback, it will also raise an `UnhandledPromiseRejectionWarning` as well? I'd like to try and get a minimum reproducible example of this, if just so I can better understand typing/TypeScript that much better.

**Comment by MasterOdin** [comment]
Thinking about this quickly, another alternative (better maybe than doing promise trickery) would be doing inline function overriding in the class:
```

function bar(data: string, callback: (err: Error | null, response?: string) => void) {
    callback(null, data);
}

class Foo {
    foo(data: string): Promise<string>;
    foo(data: string, callback?: (err: Error | null, response?: string) => void): any;

    foo(data: string, callback?: (err: Error | null, response?: string) => void): any {
        if (!callback) {
            return new Promise((resolve, reject) => {
                this.foo(data, (err, response) => {
                    err ? reject(err) : resolve(response);
                });
            })
        }

        bar(data, callback);
    }
}

const a = new Foo();
a.foo('string1').then((data) => {
    console.log(data);
})

a.foo('string2', (err, data) => {
    console.log(data);
})
```
Unfortunately, this is a bit more complicated than doing function overrides in the `*.d.ts` files as the return type for each override here has to have a compatible return type, but much like in the suggested Promise return use-case (and currently), I doubt that consumers using the callback interface would notice/care about the `any` return type.

**Comment by dpopp07** [comment]
Hmm this is a nice idea. I tried overloading methods initially and couldn't get the compiler to accept it but looking at this, my syntax might have been wrong. I'll play around with this

**Comment by evanshortiss** [comment]
@dpopp07 ah right, I didn't think about that in the initial method overloading suggestion. I think the correct solution is only to extend the example I gave to return a Promise when necessary using the suggestion by @MasterOdin 👍 

It's hairy since you need need to define method A, B, and then implement `A | B` as the return type making the callback optional per the example by @MasterOdin

**Comment by dpopp07** [comment]
@evanshortiss @MasterOdin Well after all of this, it occurred to me that `axios` (the request library we use under the hood) returns a Promise and we actually execute the callback inside of the handler for said Promise. This means that the current SDK is already broken as far as allowing Promises rejections for callback users and I'm not sure there's a way to fix it short of swapping out the library.

Since this is the case, we are thinking that we'll probably go back to the original solution of always returning a Promise and at least improve that experience. We'll probably declare callbacks as deprecated in the near future and move to only Promises in the later future. Let me know if you have other ideas but I'm going to start implementing this today.

**Comment by MasterOdin** [comment]
Would it be possible to highlight the entry point into axios and then were its generally throwing an exception from, messing up the callback chain?

**Comment by dpopp07** [comment]
Promises now have a clearly defined resolution type that allows inferred typing, starting in v5.0.0
